26 implement substitution alternative

znittzel opened the pull request.

znittzel committed:
Top-down easy reducing method implemented

znittzel committed:
Tests added and small changes implemented

znittzel committed:
Tests ok

ourmoa commented on src/lib.rs:
```diff
@@ -432,6 +443,12 @@ impl Clone for Variable {
     }
 }
 
+impl PartialEq for Variable {
```
More descriptive function name? 
```suggestion
impl IdEq for Variable {
```

znittzel replied:
Isn't PartialEq a Trait?

ourmoa commented on src/lib.rs:
```diff
@@ -620,48 +714,563 @@ impl Theory {
+        // Depth 3
+        // 0: 1 & 2
+        // 1: 3 + 4 >= 6 (3 has bounds (-5,5), 4 is bool) 
+        // 2: 5 + 6 >= 5 (5 has bounds (-3,3), 6 is bool) 
```
```suggestion
        // 2: 5 + 6 >= 4 (5 has bounds (-3,3), 6 is bool) 
```

znittzel committed:
Documented new AtLeast functions
Testing inequalities from all variable combinations
Small restructuring in lineq_reduced

znittzel committed:
Removed example for to_lineq_extended since it should
not be public
Fixed last tests

znittzel committed:
Update src/lib.rs

Co-authored-by: ourmoa <100134345+ourmoa@users.noreply.github.com>

ourmoa commented on src/lib.rs:
```diff
@@ -6,7 +6,9 @@ use std::hash::Hash;
 use std::{collections::HashMap};
 use std::fmt::Display;
 use itertools::iproduct;
```
```suggestion
```

ourmoa commented on src/lib.rs:
```diff
@@ -6,7 +6,9 @@ use std::hash::Hash;
 use std::{collections::HashMap};
 use std::fmt::Display;
 use itertools::iproduct;
+use std::ops::Range;
```
```suggestion
```

ourmoa commented on src/lib.rs:
```diff
@@ -6,7 +6,9 @@ use std::hash::Hash;
 use itertools::iproduct;
+use std::ops::Range;
 use std::cmp;
+use itertools::Itertools;
```
```suggestion
```

ourmoa commented on src/lib.rs:
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
+    }
+
+    /// Transforms into a linear inequality constraint. If sub constraints fulfills certain requirements, then the sub constraints
+    /// will be merged into this constraint. If requirements is not met, then `to_lineq_extended` is called. 
```
```suggestion
    /// will be merged into this constraint. If the requirements are not met, then `to_lineq_extended` is called. 
```

ourmoa commented on src/lib.rs:
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
-}
+}
+
+/// A `Statement` is a declaration of a expression (or proposition) connected to a `Variable`.
```
```suggestion
/// A `Statement` is a declaration of an expression (or proposition) connected to a `Variable`.
```

ourmoa commented on src/lib.rs:
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
+        return id_curr;
+    }
+
+    // All bottom / atomic variable ids. E.i. statements that don't have any children.
```
```suggestion
    // All bottom / atomic variable ids, e.i. statements that don't have any children.
```

ourmoa commented on src/lib.rs:
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
+    use super::*;
+
+    // Validates all combinations generated from `variable_bounds` in two vectors, lineqs0 och lineqs1, of GeLineq data types. 
+    // `all_relation` how the results from lineqs0 and lineqs1 relates. If `all_relation` is true then either both 
```
feels like something is missing in the last sentence?

ourmoa commented on src/lib.rs:
```diff
@@ -470,200 +469,891 @@ pub struct AtLeast {
+                indices: vec![3,4,5,6]
+            },
+        ];
+        assert!(validate_theory_lineqs(t, actual, expected));
```
I believe we want to validate the theory based on the input and actual and not on the expected

znittzel committed:
Moved imports and implementations for tests
to mod tests
Exhaustive combination testing for
 - test_theory_to_lineqs_reduced
 - test_merge_conj_disj
 - test_substitution
 - test_min_max_coefficients

znittzel committed:
Update src/lib.rs

Co-authored-by: ourmoa <100134345+ourmoa@users.noreply.github.com>

znittzel committed:
Update src/lib.rs

Co-authored-by: ourmoa <100134345+ourmoa@users.noreply.github.com>

znittzel committed:
Update src/lib.rs

Co-authored-by: ourmoa <100134345+ourmoa@users.noreply.github.com>

ourmoa commented on src/lib.rs:
````diff
@@ -470,200 +450,820 @@ pub struct AtLeast {
+
+    /// Transforms into a linear inequality constraint. Unlike `to_lineq` this is extended to include the given variable
+    /// such that it holds logic as "if `given_id` is true then `self` must be true".
+    /// ```
````
```suggestion
```
Remove?

ourmoa commented on src/lib.rs:
```diff
@@ -470,200 +450,820 @@ pub struct AtLeast {
+        }
+    }
+
+    // Validates all combinations generated from `variable_bounds` in two vectors, lineqs0 och lineqs1, of GeLineq data types. 
```
```suggestion
    // Validates that all combinations, generated from `variable_bounds`, in two vectors of GeLineqs gives the same result, i.e. they hold the same truth table```

ourmoa commented on src/lib.rs:
```diff
@@ -470,200 +450,820 @@ pub struct AtLeast {
+    }
+
+    // Validates all combinations generated from `variable_bounds` in two vectors, lineqs0 och lineqs1, of GeLineq data types. 
+    // `all_relation` how the results from lineqs0 and lineqs1 relates. If `all_relation` is true then either both 
```
```suggestion
    // The `all_relation` states how the GeLineq within each vector relates. If `all_relation` is true then all GeLineqs of the vector must hold to produce `True` as output. If `all_relation` is false then it is treated as an any, and only one GeLineq of the vector must hold to produce `True` as output. 
```

znittzel committed:
Update src/lib.rs

Co-authored-by: ourmoa <100134345+ourmoa@users.noreply.github.com>

znittzel committed:
Update src/lib.rs

Co-authored-by: ourmoa <100134345+ourmoa@users.noreply.github.com>

znittzel committed:
Update src/lib.rs

Co-authored-by: ourmoa <100134345+ourmoa@users.noreply.github.com>

znittzel committed:
Testing that actual eq expected for test_theory_to_lineqs
